### PR TITLE
feat(core): add engine computing the circuit bootstrap + vertical packing

### PR DESCRIPTION
**Pull request by IceTDrinker**
### Resolves:

refs https://github.com/zama-ai/concrete-core-internal/issues/224
refs https://github.com/zama-ai/concrete-core-internal/issues/225

### Description

- add spec and impl for LweCiphertextVectorDiscardingCircuitBootstrapBooleanVerticalPackingEngine in the fftw backend

### Checklist 

(Use '[x]' to check the checkboxes, or submit the PR and then click the checkboxes)

* [x] Tests for the changes have been added (for bug fixes / features)
* [x] Docs have been added / updated (for bug fixes / features)
* [x] The PR description links to the related issue (to link an issue, use '#XXX'.)
* [x] The tests on AWS have been launched and are successful (comment with @slab-ci cpu_test and/or @slab-ci gpu_test to trigger the tests)
* [x] The draft release description has been updated
* [x] Check for breaking changes (including serialization changes) and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

### CODE REVIEW

**Comment by IceTDrinker** [comment]
@slab-ci cpu_test

File: concrete-core/src/specification/engines/lwe_ciphertext_vector_discarding_circuit_bootstrap_boolean_vertical_packing.rs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+}
+
+impl<EngineError: std::error::Error>
+    LweCiphertextVectorDiscardingCircuitBootstrapBooleanVerticalPackingError<EngineError>
+{
+    /// Validates the inputs
+    pub fn perform_generic_checks(
+        cbs_decomposition_level_count: DecompositionLevelCount,
+        cbs_decomposition_base_log: DecompositionBaseLog,
+        integer_precision: usize,
```
Comment (agnesLeroy) [inline concrete-core/src/specification/engines/lwe_ciphertext_vector_discarding_circuit_bootstrap_boolean_vertical_packing.rs:27]: integer_precision -> ciphertext_modulus_log?

File: concrete-core/src/specification/engines/lwe_ciphertext_vector_discarding_circuit_bootstrap_boolean_vertical_packing.rs [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        }
+        if cbs_decomposition_base_log.0 * cbs_decomposition_level_count.0 > integer_precision {
+            return Err(Self::DecompositionTooLarge);
+        }
+        Ok(())
+    }
+}
+
+/// A trait for engines performing a (discarding) boolean circuit bootstrapping followed by a
+/// vertical packing on LWE ciphertext vectors.
```
Comment (agnesLeroy) [inline concrete-core/src/specification/engines/lwe_ciphertext_vector_discarding_circuit_bootstrap_boolean_vertical_packing.rs:43]: Maybe we could add: The term "boolean" refers to the fact the input ciphertexts encrypt single bits of message.

File: concrete-core/src/specification/engines/lwe_ciphertext_vector_discarding_circuit_bootstrap_boolean_vertical_packing.rs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+/// vector. It then fills the `output` LWE ciphertext vector with the result of the vertical packing
+/// operation applied on the output of the circuit bootstrapping, using the provided look-up table.
+///
+/// # Formal Definition
+pub trait LweCiphertextVectorDiscardingCircuitBootstrapBooleanVerticalPackingEngine<
+    Input,
+    Output,
+    BootstrapKey,
+    LUT,
+    CirctuiBootstrapFunctionalPackingKeyswitchKeys,
```
Comment (agnesLeroy) [inline concrete-core/src/specification/engines/lwe_ciphertext_vector_discarding_circuit_bootstrap_boolean_vertical_packing.rs:58]: Typo in CircuitBootstrapFunctionalPackingKeyswitchKeys (Circtui)

File: concrete-core/src/backends/fftw/implementation/engines/fftw_engine/lwe_ciphertext_vector_discarding_circuit_bootstrap_boolean_vertical_packing.rs [hunk 4]
```diff
@@ -0,0 +1,481 @@
...
+    /// let polynomial_size = PolynomialSize(1024);
+    /// let glwe_dimension = GlweDimension(1);
+    /// let lwe_dimension = LweDimension(481);
+    ///
+    /// let var_small = Variance::from_variance(
+    ///     StandardDev::from_standard_dev(0.000_000_000_000_000_221_486_881_160_055_68).get_variance(),
+    /// );
+    /// let var_big = Variance::from_variance(
+    ///     StandardDev::from_standard_dev(0.000_061_200_133_780_220_36).get_variance(),
+    /// );
```
Comment (agnesLeroy) [inline concrete-core/src/backends/fftw/implementation/engines/fftw_engine/lwe_ciphertext_vector_discarding_circuit_bootstrap_boolean_vertical_packing.rs:53]: Here using 10^-100 values would look more like what we do in other doc tests...

**Review by agnesLeroy — APPROVED** [review]
Thanks a lot for this! I just have two small comments you'll see. :slightly_smiling_face:

File: concrete-core/src/specification/engines/lwe_ciphertext_vector_discarding_circuit_bootstrap_boolean_vertical_packing.rs (hunk 1 above)
Comment (IceTDrinker) [reply concrete-core/src/specification/engines/lwe_ciphertext_vector_discarding_circuit_bootstrap_boolean_vertical_packing.rs:27]: I re-used the integer_precision from somewhere else, so we must have the two wordings in the code base -_-

File: concrete-core/src/backends/fftw/implementation/engines/fftw_engine/lwe_ciphertext_vector_discarding_circuit_bootstrap_boolean_vertical_packing.rs (hunk 4 above)
Comment (IceTDrinker) [reply concrete-core/src/backends/fftw/implementation/engines/fftw_engine/lwe_ciphertext_vector_discarding_circuit_bootstrap_boolean_vertical_packing.rs:53]: yep will do

File: concrete-core/src/specification/engines/lwe_ciphertext_vector_discarding_circuit_bootstrap_boolean_vertical_packing.rs (hunk 1 above)
Comment (IceTDrinker) [reply concrete-core/src/specification/engines/lwe_ciphertext_vector_discarding_circuit_bootstrap_boolean_vertical_packing.rs:27]: In any case, I'll create a check for the decomposition stuff as it's duplicated all over the place and driving me crazy 😅 

btw should we have a time having both information ?

File: concrete-core/src/backends/fftw/implementation/engines/fftw_engine/lwe_ciphertext_vector_discarding_circuit_bootstrap_boolean_vertical_packing.rs (hunk 4 above)
Comment (IceTDrinker) [reply concrete-core/src/backends/fftw/implementation/engines/fftw_engine/lwe_ciphertext_vector_discarding_circuit_bootstrap_boolean_vertical_packing.rs:53]: done locally

File: concrete-core/src/specification/engines/lwe_ciphertext_vector_discarding_circuit_bootstrap_boolean_vertical_packing.rs (hunk 3 above)
Comment (IceTDrinker) [reply concrete-core/src/specification/engines/lwe_ciphertext_vector_discarding_circuit_bootstrap_boolean_vertical_packing.rs:58]: done locally

**Comment by IceTDrinker** [comment]
@slab-ci cpu_test

**Pull request merged**